### PR TITLE
Schema: Tests, minor refactor and one or two small bugfixes

### PR DESCRIPTION
**Pull request by ewels**
Turns out that if you leave the command line tool waiting for ages, it eventually dies in an ugly way with a `RecursionError`:

```
RecursionError: maximum recursion depth exceeded while calling a Python object
```

This PR just catches those exceptions and exits nicely.

_Edit:_ Also scooped into this PR:

* Template update to make font awesome icon just the class names
* Loads of new tests for `schema.py`
* Some code refactoring to reduce duplication

### CODE REVIEW

**Comment by ewels** [comment]
Note that I probably _should_ refactor the code so that the function doesn't call itself but instead uses a loop or something to avoid this. But it takes ages to reach this recursion limit so I think it's probably fine as it is.

**Comment by ewels** [comment]
So close to getting CodeCov to pass!

> _60% of diff hit (target 62.14%)_

**Comment by ewels** [comment]
All tests passing! ✅ 

Ready for merge if anyone fancies giving a review 😄

File: nf_core/schema.py [hunk 1]
```diff
@@ -163,56 +166,42 @@ def build_schema(self, pipeline_dir, no_prompts, web_only, url):
...
-                "properties": {}
-            }
+            self.make_skeleton_schema()
+
+        # Load and validate Schema
+        try:
+            self.lint_schema()
+        except AssertionError as e:
+            logging.error("Existing JSON Schema found, but it is invalid: {}".format(click.style(str(self.schema_filename), fg='red')))
+            logging.info("Please fix or delete this file, then try again.")
```
Comment (mashehu) [inline nf_core/schema.py]: I don't know how good the schema linting is, but could we return fields which should be fixed in this message?

Comment (ewels) [reply nf_core/schema.py]: Yup, the linting already logs error messages itself so these will be printed above this message.

File: tests/test_schema.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    def test_validate_schema_fail_notjsonschema(self):
+        """ Check that the schema validation fails when not JSONSchema """
+        self.schema_obj.schema = {'type': 'invalidthing'}
+        self.schema_obj.validate_schema()
+
+    @pytest.mark.xfail(raises=AssertionError)
+    def test_validate_schema_fail_nfcore(self):
+        """ Check that the schema validation fails nf-core addons """
+        self.schema_obj.schema = {}
+        self.schema_obj.validate_schema()
```
Comment (mashehu) [inline tests/test_schema.py]: isn't that only a test for a failed validation of an empty schema?

File: tests/test_schema.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        self.schema_obj.get_wf_params()
+
+    def test_prompt_remove_schema_notfound_config_returntrue(self):
+        """ Remove unrecognised params from the schema """
+        self.schema_obj.pipeline_params = {'foo': 'bar'}
+        self.schema_obj.no_prompts = True
+        assert self.schema_obj.prompt_remove_schema_notfound_config('baz')
+
+    def test_prompt_remove_schema_notfound_config_returnfalse(self):
+        """ Do not temove unrecognised params from the schema """
```
Comment (mashehu) [inline tests/test_schema.py]: typo "temove"

File: tests/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    def test_launch_web_builder_connection_error(self, mock_post):
+        """ Mock launching the web builder, but get a connection error """
+        # Define the behaviour of the request get mock
+        mock_post.side_effect = requests.exceptions.ConnectionError()
+        self.schema_obj.launch_web_builder()
+
+    @pytest.mark.xfail(raises=AssertionError)
+    @mock.patch('requests.post')
+    def test_get_web_builder_response_timeout(self, mock_post):
+        """ Mock chekcing for a web builder response, but timeout on the request """
```
Comment (mashehu) [inline tests/test_schema.py]: typo "chekcing"

**Comment by mashehu** [comment]
Looks good, just some small typos and one test, I didn't completely understand.

File: tests/test_schema.py (hunk 2 above)
Comment (ewels) [reply tests/test_schema.py]: Yup - although I think an empty JSON object is probably valid JSON Schema, we test to make sure that it has at least a `properties` key. So this should raise that nf-core specific `AssertionError`

**Comment by ewels** [comment]
Thanks! Typo fixed and added to the docstring for the test that was confusing. Hopefully that makes more sense now.

**Pull request merged**